In [1]:
%pip install pdfminer.six > nul
%pip install rich > nul
%pip install tqdm > nul
%pip install nltk > nul

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [2]:
%pip install  langchain-pinecone > nul
%pip install  sentence-transformers > nul
%pip install  langchain-community > nul
%pip install  PyPDF2 > nul
%pip install  rouge > nul

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-cloud-bigquery 2.34.4 requires packaging<22.0dev,>=14.3, but you have packaging 24.2 which is incompatible.
jupyterlab 4.2.5 requires jupyter-lsp>=2.0.0, but you have jupyter-lsp 1.5.1 which is incompatible.
jupyterlab-lsp 5.1.0 requires jupyter-lsp>=2.0.0, but you have jupyter-lsp 1.5.1 which is incompatible.
kfp 2.5.0 requires google-cloud-storage<3,>=2.2.1, but you have google-cloud-storage 1.44.0 which is incompatible.
kfp 2.5.0 requires requests-toolbelt<1,>=0.8.0, but you have requests-toolbelt 1.0.0 which is incompatible.
libpysal 4.9.2 requires shapely>=2.0.1, but you have shapely 1.8.5.post1 which is incompatible.
thinc 8.3.2 requires numpy<2.1.0,>=2.0.0; python_version >= "3.9", but you have numpy 1.26.4 which is incompatible.
ydata-profiling 4.10.0 requires scipy<1.14,>=1.4.1, but you have scipy 

In [3]:
# Libs for inner connection
import os
import nltk
nltk.download('punkt')  # need to tokenize
from tqdm import tqdm
import re
import pandas as pd 
import numpy as np
from rich import print

# Libs for preprocessing text
from nltk.tokenize import sent_tokenize
from PyPDF2 import PdfReader
from transformers import AutoTokenizer, AutoModel, pipeline

# Libs to vectorize data and make dataspace?
import torch
from pinecone import Pinecone, ServerlessSpec
from langchain.embeddings.huggingface import HuggingFaceEmbeddings
from uuid import uuid4
import time

# Libs for ChatBot
from langchain_pinecone import PineconeVectorStore
from langchain.chains.conversation.memory import ConversationBufferWindowMemory
from langchain.llms import HuggingFaceHub
from langchain.chains import RetrievalQA

#Libs for model.eval
import json
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rouge import Rouge

[nltk_data] Downloading package punkt to /usr/share/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


# Preprocessing PDF

In [4]:
reader = PdfReader("/kaggle/input/doc-pdf/causeway_bbva_classic_autocallable_3.pdf")

In [66]:
def doc_cleaner(reader):
    '''
    ----------------------------------
    The func that deleted useless content such, as headers, endings, 
    hieroglyphs, e.c.t.
    ----------------------------------
    Parameters:
    reader - class for extractung text from PDF
    ----------------------------------
    '''

    clean_pages = []
    
    for page in reader.pages:
    
        str_page = page.extract_text()
        str_page = re.sub( r".*? \n BBVA  \nCLASSIC AUTOCALLABLE NOTE 22  \nNovember 2024", "", str_page, flags=re.MULTILINE) # удаляю header страницы
        str_page = re.sub(r".*?www.causeway -securities.com\s+", "", str_page, flags=re.DOTALL) # удаление ending
        str_page = re.sub(r"\s*\n\n•\s*", " ", str_page, flags=re.DOTALL)
        str_page = re.sub(r"\s*\n\n", " ", str_page, flags=re.DOTALL)
        str_page = re.sub(r"\s*\n", " ", str_page, flags=re.DOTALL)
        str_page = re.sub(r"  ", " ", str_page, flags=re.DOTALL)
        clean_pages.append(str_page.strip())

    return clean_pages

In [67]:
clear_text = '\n'.join(doc_cleaner(reader))
len(clear_text)

13128

In [69]:
def split_into_sentence_chunks(text, max_chunk_length):
  '''
  ----------------------------------
  The func that split the solid text into chunks, using 
  sent_tokenize splitter and checking the length accordance
  ----------------------------------
  Parameters:
  text - the solid text that should be chunked (type=str)
  max_chunk_length - the length of each chunck (type=int)
  ----------------------------------
  '''
  sentences = sent_tokenize(text)
    
  current_chunk = ""
  chunks = []

  for sentence in sentences:
      if len(current_chunk) + len(sentence) > max_chunk_length:
          chunks.append(current_chunk)
          current_chunk = sentence
      else:
          current_chunk += sentence

  if current_chunk:
      chunks.append(current_chunk)

  return chunks

In [113]:
max_chunk_length = 100
sentence_chunks = split_into_sentence_chunks(clear_text, max_chunk_length)

In [114]:
df = pd.DataFrame(sentence_chunks, columns=['content'])
print(df.shape)
df.head()

(85, 1)

,content
0,
1,CCY ISIN Investment Return Maximum Term Autoca...
2,6 Years Semi -Annual from end of Y1 100% 65% G...
3,Strike Date: 22 November 2024 Issue Date: 29 N...
4,"The maximum product term is 6 years, with mult..."


In [115]:
# Loading the model and it's tokenizer
model_name = "bert-base-cased"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name)

def embed_text(text):
    '''
    ----------------------------------
    The func that uses Bert tokinazer to embed the sentences
    ----------------------------------
    Parameters:
    text - the cleaned text that should be chunked (type=str)
    ----------------------------------
    '''
    
    inputs = tokenizer(text, return_tensors="pt", truncation=True, padding=True, max_length=512)
    with torch.no_grad():
        outputs = model(**inputs)
    # Taking the embedding of CLS token 
    return outputs.last_hidden_state[:, 0, :].squeeze().numpy()

df["embedding"] = df["content"].apply(embed_text)
print(df)


/opt/conda/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1617: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be deprecated in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


content  \
0                                                       
1   CCY ISIN Investment Return Maximum Term Autoca...   
2   6 Years Semi -Annual from end of Y1 100% 65% G...   
3   Strike Date: 22 November 2024 Issue Date: 29 N...   
4   The maximum product term is 6 years, with mult...   
..                                                ...   
80                               All rights reserved.   
81  No part of this publication may be reproduced,...   
82  The securities described in this document have...   
83  The securities described in this document may ...   
84  Person (as defined in Regulation S under the S...   

                                            embedding  
0   [0.33139247, 0.23923527, 0.25130033, -0.736903...  
1   [0.44682628, 0.07769964, 0.1091394, -0.2152123...  
2   [0.4631762, 0.11358263, 0.11897047, -0.0657125...  
3   [0.34657794, -0.05354807, -0.023549935, -0.193...  
4   [0.3857128, -0.030409144, 0.019222502, -0.0386...  
..                                                ...  
80  [-0.0031022362, -0.22165984, 0.32388386, -0.51...  
81  [-0.11940248, -0.29665416, 0.0781199, -0.48900...  
82  [0.38853127, -0.22295335, 0.03471762, -0.36565...  
83  [0.44091478, -0.16944173, -0.09550869, -0.3037...  
84  [0.4176536, -0.12064551, 0.16685916, -0.679027...  

[85 rows x 2 columns]

In [116]:
# Saving locally for later
data_path = "/kaggle/working//parsed_pdf_doc_with_embeddings.csv"
df.to_csv(data_path, index=False)

# RAG

In [117]:
index_name = 'pdfsearch-1-0'

In [118]:
pc = Pinecone(
    api_key="pcsk_7AbZoC_TT8fAugtqUngo18ELJ9DyRJ6AncsLs5h5R8iuxh9BBzBt15ALR5Dde6wbw3s5XK"
    )

if index_name not in pc.list_indexes():
    # create a new index
    pc.create_index(
        spec=ServerlessSpec(
            cloud='aws',
            region='us-east-1'
        ),
        name=index_name,
        metric='cosine',
        dimension=1024
    )

PineconeApiException: (409)
Reason: Conflict
HTTP response headers: HTTPHeaderDict({'content-type': 'text/plain; charset=utf-8', 'access-control-allow-origin': '*', 'vary': 'origin,access-control-request-method,access-control-request-headers', 'access-control-expose-headers': '*', 'x-pinecone-api-version': '2024-07', 'X-Cloud-Trace-Context': '633cdab11d31f3e966d383fcbe301915', 'Date': 'Sat, 30 Nov 2024 18:39:58 GMT', 'Server': 'Google Frontend', 'Content-Length': '85', 'Via': '1.1 google', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'})
HTTP response body: {"error":{"code":"ALREADY_EXISTS","message":"Resource  already exists"},"status":409}


In [119]:
pc.describe_index(index_name)

{'deletion_protection': 'disabled',
 'dimension': 1024,
 'host': 'pdfsearch-1-0-7htp840.svc.aped-4627-b74a.pinecone.io',
 'metric': 'cosine',
 'name': 'pdfsearch-1-0',
 'spec': {'serverless': {'cloud': 'aws', 'region': 'us-east-1'}},
 'status': {'ready': True, 'state': 'Ready'}}

In [120]:
embed = HuggingFaceEmbeddings(model_name="sentence-transformers/all-roberta-large-v1")

/opt/conda/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1617: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be deprecated in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


In [121]:
index = pc.Index(index_name)

batch_size = 4
for i in tqdm(range(0, len(sentence_chunks), batch_size)):
    
    i_min = min(i+batch_size, len(sentence_chunks))
    batch = sentence_chunks[i: i_min]
    meta_data = [{"titile" : 'BBVA CLASSIC AUTOCALLABLE NOTE 22', 
              "context": row}
                for row in batch]
    ids = [str(uuid4()) for _ in range(len(batch))]
    # Encode the text to obtain its vector representation
    embeds = embed.embed_documents(batch)
    
    # Upsert the vector and text into the Pinecone index
    index.upsert(vectors=zip(ids, embeds, meta_data))
    time.sleep(4)

100%|██████████| 22/22 [02:00<00:00,  5.48s/it]


# Prompt

In [122]:
text_field = 'context'

vectorstore = PineconeVectorStore(
    index, embed, text_field
)

In [123]:
query = "When is the Final Valuation Date?"

vectorstore.similarity_search(
    query,  # our search query
    k=1  # return 1 most relevant info vector from vectorstore
)

[Document(id='2a5d2efc-b3af-4bde-85bd-86b917c2fa7f', metadata={'titile': 'BBVA CLASSIC AUTOCALLABLE NOTE 22'}, page_content='OBSERVATION DATES Observation Date Payment Date Autocall Barrier Observation 1 - - - Observation 2 25 November 2025 3 December 2025 100% Observation 3 22 May 2026 1 June 2026 100% Observation 4 24 November 2026 2 December 2026 100% Observation 5 24 May 2027 1 June 2027 100% Observation 6 22 November 2027 30 November 2027 100% Observation 7 22 May 2028 30 May 2028 100% Observation 8 22 November 2028 30 November 2028 100% Observation 9 22 May 2029  30 May 2029 100% Observation 10 26 November 2029 3 December 2029 100% Observation 11 22 May 2030  30 May 2030 100% Final Valuation Date 22 November 2030 2 December 2030 100% (65% Protection Barrier) *Some dates may vary slightly depending on bank holidays or days that Underlyings do not trade.')]

# ChatBot

In [143]:
llm = HuggingFaceHub(
    repo_id="google/flan-t5-large", # the most acceptable model, other showed the worth result
    model_kwargs={"temperature": 0.3, "max_length": 512},
    huggingfacehub_api_token='hf_gyWbZsWiEWXzKtestJWaBzIXdTSWILGPCl'
)

# conversational memory
conv_mem = ConversationBufferWindowMemory(
    memory_key = 'chat_history',
    k = 5,
    return_messages =True)

# retrieval qa
qa = RetrievalQA.from_chain_type(
    llm = llm,
    chain_type = "stuff",
    retriever = vectorstore.as_retriever(), 
    memory=conv_mem
)

In [144]:
query = "What is the final valuation date for this product?"
qa.invoke(query)

{'query': 'What is the final valuation date for this product?',
 'chat_history': [],
 'result': '22 November 2030 2 December 2030'}

In [111]:
qa_list = [
    {"question": "Who is the issuer of this product?", "answer": "BBVA Global Markets B.V. (A3 / A)"},
    {"question": "What is the maximum term of the product?", "answer": "6 years"},
    {"question": "How often are autocall observations made?", "answer": "semi-annual Autocall Observation date from end of year 1"},
    {"question": "What is the capital protection in this product?", "answer": "Capital is at risk if any of the Underlyings are below the Protection Barrier at the Final Valuation Date or if the product issuer defaults."},
    {"question": "What is the protection barrier for each underlying?", "answer": "65% Protection Barrier"},
    {"question": "What is the annual return for USD?", "answer": "11.0% p.a."},
    {"question": "What is the annual return for GBP?", "answer": "10.5% p.a."},
    {"question": "When does the observation period for the first autocall begin?", "answer": "Observation 2 25 November 2025"},
    {"question": "How is the return calculated if all Underlyings are at or above the Autocall Barrier?", "answer": "The product will return 100% of invested capital plus an investment return of 5.50% USD or 5.25% GBP for each 6-month period that has elapsed since the Strike Date."},
    {"question": "What happens if the Underlyings are below the Protection Barrier at maturity?", "answer": "Invested capital will be reduced in line with the performance of the worst performing Underlying."},
    {"question": "What is the final redemption return for USD?", "answer": "100% of invested capital plus an investment return of 66.0% USD."},
    {"question": "What is the final redemption return for GBP?", "answer": "100% of invested capital plus an investment return of 63.0% GBP."},
    {"question": "What is the maximum potential capital loss?", "answer": "If any of the Underlyings are below the Protection Barrier, the capital repayment will be decreased in line with the performance of the worst Underlying."},
    {"question": "What is the risk related to issuer default?", "answer": "Risk of partial or total loss of capital and no income in the case of bankruptcy or payment default by the Issuer."},
    {"question": "What is the capital risk if the product is sold early?", "answer": "Selling out of note early may result in a capital loss."},
    {"question": "How is the performance of the product affected by volatility?", "answer": "The price will depend on numerous factors, including the level of volatility of the underlying indices, the remaining time to maturity, interest rates, and the perception of the Issuer's credit quality."},
    {"question": "What is the coupon memory feature?", "answer": "The Coupon Memory feature allows previously missed coupons to be recaptured."},
    {"question": "What are the conditions for early redemption?", "answer": "If all the Underlyings are at or above the Autocall Barrier on any semi-annual Autocall Observation date from the end of Year 1."},
    {"question": "What happens if the Underlyings close at or above the Autocall Barrier at maturity?", "answer": "The product will return 100% of invested capital plus an investment return of 66.0% USD or 63.0% GBP."},
    {"question": "What are the indices used as underlyings for this product?", "answer": "S&P 500 Index (SPX), Nikkei 225 Index (NKY), EURO STOXX 50 Index (SX5E)."},
    {"question": "What is the risk of income potential in this product?", "answer": "Income potential is capped since investors do not participate directly in any capital growth in the Underlyings."},
    {"question": "What is the S&P 500 Index?", "answer": "The S&P 500 is widely regarded as the best single gauge of large-cap U.S. equities and serves as the foundation for a wide range of investment products."},
    {"question": "What is the Nikkei 225 Index?", "answer": "The Nikkei-225 Stock Average is a price-weighted average of 225 top-rated Japanese companies listed in the First Section of the Tokyo Stock Exchange."},
    {"question": "What is the EURO STOXX 50 Index?", "answer": "The EUROSTOXX 50 Index, Europe’s leading blue-chip index for the Eurozone, provides a blue-chip representation of supersector leaders in the Eurozone."},
    {"question": "What is the credit rating of BBVA?", "answer": "A3 by Moody’s, A by S&P."},
    {"question": "What are the suitability requirements for this product?", "answer": "Investors should have received professional financial advice, understand financial markets and structured notes, seek capital growth, and be prepared to invest for the medium to long-term."},
    {"question": "What is the risk related to issuer bankruptcy?", "answer": "If the Issuer defaults, investors could lose some or all of their invested capital and no returns."},
    {"question": "What is the role of Causeway Securities?", "answer": "Causeway Securities is an independent cross-asset brokerage authorized in the UK by the FCA, offering personalized and independent services for structured investment solutions."},
    {"question": "What is the role of BBVA in the product?", "answer": "BBVA is the issuer of the product and provides investment banking services, asset management, and securities brokerage services."},
    {"question": "What happens if the investor sells the note early?", "answer": "The price will depend on factors such as volatility, time to maturity, interest rates, and the credit quality of the issuer, and the investor may receive less than the initially invested amount."},
    {"question": "What is the final valuation date for this product?", "answer": "22 November 2030"},
    {"question": "What is the observation date for 2026?", "answer": "22 May"},
    {"question": "What happens if the product is redeemed early?", "answer": "Investors receive 100% of invested capital plus the applicable coupon return."},
    {"question": "What is the protection barrier at maturity?", "answer": "If all the Underlyings are at or above the Protection Barrier, 100% of invested capital is returned."},
    {"question": "What is the risk of investing in this product?", "answer": "Investors face the risk of partial or total loss of capital if the Underlyings are below the Protection Barrier at maturity or if the Issuer defaults."}
]


# Evaulation

In [145]:
def evaluate_model(qa, questions_and_answers):

    '''
    ----------------------------------
    The func that uses Bert tokinazer to embed the sentences
    ----------------------------------
    Parameters:
    qa - question-answering class with the retriever and LLM model (type=RetrievalQA:object)
    questions_and_answers - the list containing dicts with client's questions about project 
    ----------------------------------
    '''
    
    predicted_answers = [qa.invoke(item["question"]) for item in questions_and_answers]
    ground_truth_answers = [item["answers"] for item in questions_and_answers]

    # Checking for the string type
    predicted_answers = [str(answer) for answer in predicted_answers]
    ground_truth_answers_flattened = [" ".join(answers) for answers in ground_truth_answers]

    # BLEU Score
    smoothing_function = SmoothingFunction().method4
    bleu_scores = [
        sentence_bleu([true.lower().split()], pred.lower().split(), smoothing_function=smoothing_function)
        for true, pred in zip(ground_truth_answers_flattened, predicted_answers)
    ]
    average_bleu_score = sum(bleu_scores) / len(bleu_scores)

    # ROUGE Score
    rouge = Rouge()
    rouge_scores = rouge.get_scores(predicted_answers, ground_truth_answers_flattened, avg=True)

    return average_bleu_score, rouge_scores

# Оценка модели
average_bleu_score, rouge_scores = evaluate_model(qa, questions_and_answers)
print(f"Average BLEU Score: {average_bleu_score}")
print(f"ROUGE Scores: {rouge_scores}")

Average BLEU Score: 0.09167034103051429

ROUGE Scores: {'rouge-1': {'r': 0.3523142755191436, 'p': 0.15730431006134096, 'f': 0.2118116400464035}, 'rouge-2': 
{'r': 0.11152197540364435, 'p': 0.04573968768977015, 'f': 0.06356957188284919}, 'rouge-l': {'r': 
0.3125177441134169, 'p': 0.14221036752553973, 'f': 0.19004124811831058}}

# Summary  


В данной задаче была реализована система question-answer на основе ChatBot с подключением API, ниже будут разобраны результаты работы  

* Preprocessing PDF  

  Сначала была использована сырая функция text_extract() из библиотеки, однако вытягиваемый текст из PDF почему-то обрезался (пропадали данные о датах, смешивались таблицы ...), поэтому в целях улучшения качества была использована PdfReader, она позволила ублать пользовательские функции парсинга страниц, а так же извлекла текст в полной сохранности, что позволило улучшить финальную модель  

* RAG/PROMPT  

  После реализации вектороного пространства хранения данных (контента) появились новые возможности улучшения модели (пиллинг RAG): размерность вектора; токинизатор; LLM модель; размерность обрабатываемых данных (batch) и размерность контента (chunk)  

  1) **Размерность вектора:**  были рассмотренны размерности 768; 1024; 2048. Размерность 768 позволяла модели слишком обобщать ответы в данной задачи, из-за чего они получались правильные, но слишком общие, содержащие ответ не относящийся к вопросу; 1024 оптимальная размерность подобранная мною, поскольку модель дает точные ответы на заданные вопросы (однако остались некоторые проблемы, связанные с реализацией, которые будут обозначены ниже); 2048 заставляет модель слишком досконально исследовать сходство, из-за чего модель отвечает правильно на вопрос, но добавляет еще информацию связанную с любым вхождением ключевого слова  
  2) **Токинизатор**: зависит напрямую от размерности, поскольку данные в векторном пространсве должны быть стандартизированы и согласованы, в осоновном перебор токинизатора шел из-за подбора LLM модели, для лучшей согласованности работы с embeddings, но значительных улучшений качетсва это не дало  
  3) **Batch**: так же не дает значительного улучшения качетсва, релевантный размер 4-8, сокрее влияет на скорость обработки запросов и работы с векторным пространством данных в Pinecone, заметно на больших моделях по типу `google/flan-t5-xxl`  
  4) **Размерность контента**: позволяет значительно улучшать качество, поскольку от нее зависит `сколько предложений` будет в векторе, что позволит модели находить взаимосвязь для дальнейшего поиска, однако, учитывая финальное качетсво, хорошего перекрытия не удалось достичь  
  5) **LLM**: позволяет значительно улучшить качетсво, поскольку от нее напрямую зависит генерация, было рассмотернно множество моделей google, prithivida, вариация Bert, однако по соотношению качества и затратной мощности была выбрана `google/flan-t5-large`  

* Объяснение результатов  

  1) BLEU измеряет, насколько сгенерированный текст совпадает с эталонным на уровне слов и фраз. Значение 0.09167 указывает на очень низкое качество генерации, что говорит о слабом совпадении между сгенерированным текстом и эталоном. Хотя п обольшей части согласно `https://huggingface.co/spaces/evaluate-metric/bleu` эта оценка качества относится больше к машинному переводу нежели к семантике, я ее использовал в целях возможного дальнейшего улучшения работы с запросами на других языках  
     
  2) ROUGE-1 (unigram overlap):  
     
  r (recall): 0.3523 — модель распознала около 35.23% слов, совпадающих с эталоном.  
  p (precision): 0.1573 — лишь около 15.73% слов в сгенерированном тексте релевантны.  
  f (F1): 0.2118 — общее качество на уровне слов ниже среднего.  
     
  3) ROUGE-2 (bigram overlap):  

    r: 0.1115 — только 11.15% биграмм из эталона найдены.  
    p: 0.0457 — точность на уровне биграмм низкая.  
    f: 0.0636 — F1 на уровне биграмм указывает на плохую согласованность.  

  4) ROUGE-L (Longest Common Subsequence):  

    r: 0.3125 — модель уловила около 31.25% длинных последовательностей.  
    p: 0.1422 — точность также низкая.  
    f: 0.1900 — F1 показывает, что модель не очень хорошо улавливает структуру текста.  

BLEU и ROUGE показывают, что качество текста на данный момент низкое. Модель слабо справляется с воспроизведением структуры и содержания референсных ответов.  

* Оставшиеся проблемы и возможные решения:

  1) Одна из легко детектируемых неисправностей модель путает FinalValuation Date и Maturity Date, причина не понятна, текст хорошо разделен, даты находятся равноудаленно от ключевых слов, с большой вероятностью такая проблема наблюдается и на вопросах относящихся к концу контента
  2) Недоступность к мощным моделям: нет доступа к OpenAI поскольку требуется иностранный телефон, не уверен что это помогло бы сильно исправить финальное качетсво, поскольку предложенная мноб модель не является слабой, но как 1 из вариантов решения сложившейся проблемы
  3) Возможно главная проблема кроется в самой реализации RAG и ретривере, хотя Pinecone один и самых удобных и структурированных видов векторного хранилища и структуризации данных для обучения. Можно было бы воспользоваться изначальной идеей, которую я реализовывал, а именно кастомная реализация модели основанная исключительно на данных PDF файла, работа с датафреймом, и использования ее в качетсве ретривера, однако данную идею доработать не успел
  4) Как вариант не был опробован метод распознавания текста через изображение, возможно такой метод исключит сложности и возможную нерелевантность работы с Pinecone, что приведет к значительным улучшениям
  5) Так же не были использованы дополнительные возможности Pinecone в виде фильтрации с помощью metedata, дополнительные идентификаторы разбитого контекста и прочее, что позволило бы модели с заданной заранее инструкцией prompt более точно понимать задачу и четче искать ответы на хорошо поставленные вопросы
  6) Качество модели относительно, поскольку оно зависит по большей части от самого клиента, а именно человеческого фактора в виде заданных вопросов, однако нельзя исключать проблемы указанные в пункте 1 и 3, поскольку модель не просто не дает четкого ответа на вопрос, а иногда дает неверный